feat(motion): add extended support for reduced motion

layershifter opened the pull request:
## Previous Behavior

No way to customize motions' behavior when reduced motion is enabled.

## New Behavior

- Adds `reducedMotion` option to `AtomMotion` type.
- Tests added
- Story added

By default, when [reduced motion](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion) is enabled the duration of the animation is set to `1ms` (current behavior).

`reducedMotion` allows to customize a reduced motion version of the animation:

```ts
const Motion = createPresenceComponent({
  enter: {
    keyframes: [
      { opacity: 0, transform: 'scale(0)' },
      { opacity: 1, transform: 'scale(1)' },
    ],
    duration: 300,

    /** 💡reduced motion will not have scale animation with a different duration */
    reducedMotion: {
      keyframes: [{ opacity: 0 }, { opacity: 1 }],
      duration: 1000,
    },
  },
  exit: {
    /* ... */
  },
});
```

> 💡Note, if `keyframes` are provided, they will be used instead of the regular keyframes.

This will also works with arrays of atoms as `reducedMotion` definitions are collocated to its atoms:

```ts
const Motion = createPresenceComponent({
  enter: [
    /** Atom A
        Has a definition for "reducedMotion" */
    {
      keyframes: [
        /* ... */
      ],
      reducedMotion: {
        /* some options or keyframes*/
      },
    },
    /** Atom B
        This atom _may_ declare "reducedMotion" (it's optional) */
    {
      keyframes: [
        /* ... */
      ],
    },
  ],
  exit: [
    /* ... */
  ],
});
```

## Rejected options

### Second argument to `createMotionComponent()` & `createPresenceComponent()`

> Note: examples are based on `createPresenceComponent()` as its syntax is more verbose and highlights challenges better.

We have already an array syntax for multiple atoms (see [createPresenceComponent() and arrays](https://react.fluentui.dev/?path=/docs/motion-apis-createpresencecomponent--docs#arrays)):

```js
// 💡 We already have this API implemented
createPresenceComponent({
  enter: [
    { keyframes: [] /* Atom A: enter */},
    { keyframes: [] /* Atom B: enter */},
  ],
  exit: [/* ... */],
});
```

The proposal is two have a second argument:

```js
// ⚠️ Proposed API (rejected option)
createPresenceComponent([
  {
    enter: [
      { keyframes: [] /* Atom A: enter */ },
      { keyframes: [] /* Atom B: enter */ },
    ],
    exit: [/* ... */],
  },
  {
    enter: [
      { keyframes: [] /* Atom A: enter & reduced motion */ },
      { keyframes: [] /* Atom B: enter & reduced motion */ },
    ],
    exit: [/* ... */],
  },
]);
```

However, it will be _hard_ to distinguish between atoms and reduced motion declarations - you have to know the order of arguments in the functions to understand it.

This implementation also brings a challenge of collocation: reduced options should be optional, however we will need to maintain the order to make it work. The example below shows the requirement:

```js
// ⚠️ Proposed API (rejected option)
createPresenceComponent([
  {
    enter: [
      { keyframes: [] /* Atom A: enter */ },
      { keyframes: [] /* Atom B: enter */ },
    ],
    exit: [/* ... */],
  },
  {
    enter: [
      { keyframes: [] /* Atom B: enter & reduced motion */ },
      // 💥 💥 💥 
      // There are two atoms in the motion declaration, however only one in reduced options
      // The order will break and options will be wrongly applied
    ],
    exit: [/* ... */],
  },
]);
```

Consumers will be forced to pass `null` (_or something else?_) in this case:

```js
// ⚠️ Proposed API (rejected option)
createPresenceComponent([
  {
    enter: [
      { keyframes: [] /* Atom A: enter */ },
      { keyframes: [] /* Atom B: enter */ },
    ],
    exit: [
      { keyframes: [] /* Atom A: exit */ },
      { keyframes: [] /* Atom B: enter */ },
    ],
  },
  {
    enter: [
      null, /* Atom A: enter & reduced motion */,
      { keyframes: [] /* Atom B: enter & reduced motion */ },
    ],
    exit: [/* ... */],
  },
]);
```

### Nest atoms

That's very similar to previous option, but instead we will have syntax similar to [fallback values in Griffel](https://griffel.js.org/react/api/make-styles#css-fallback-properties).

```js
// ⚠️ Proposed API (rejected option)
createPresenceComponent([
  {
    enter: [
      [
        { keyframes: [], duration: 1000 /* Atom A: enter */ },
        { keyframes: [], duration: 300 /* Atom A: enter, reduced options */ },
        // 👆 reduced options are collocated there
      ],
      [{ keyframes: [] /* Atom B: enter */ }],
    ],
    exit: [/* ... */],
  },
]);
```

In this case we collocate reduced options with the atom they are related to, so we don't need to maintain the order like in the previous option 👍

However, this option has the same issue with being implicit and hard to understand: _which array stands for atoms and which for reduced options?_

## Related issues

Fixes #33358.

fabricteam commented on packages/react-components/react-motion/library/src/contexts/MotionBehaviourContext.ts:
```diff
@@ -6,5 +6,6 @@ import * as React from 'react';
 export type MotionBehaviourType = 'skip' | 'default';
```
## 🕵🏾‍♀️ [<b data-uid='vrt-gist'></b> visual regressions to review in the fluentuiv9 Visual Regression Report](https://aka.ms/visual-changes-review?organization=uifabric&project=fabricpublic&prId=33353&commitId=63706bddd0fcea901aab8384bbb9ce6138027083&env=prod&clientType=FLUENTUI)


  <details><summary><b>Avatar Converged</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar Converged.Badge Mask RTL.chromium.png|26| Changed|
|Avatar Converged.badgeMask.chromium.png|2| Changed|

</details>


  <details><summary><b>Drawer</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Drawer.Full Overlay Dark Mode.chromium.png|2523| Changed|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

bsunderhus commented on packages/react-components/react-motion/library/src/types.ts:
```diff
@@ -1,4 +1,15 @@
+     *
+     * @see https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion
+     */
+    reducedMotion?: { keyframes?: Keyframe[] } & KeyframeEffectOptions;
```
it looks to me that `reducedMotion` is a redeclaration of `AtomMotion` 🤔, maybe it would be better to add it somewhere else to avoid repetition? or perhaps we should have a middle type here between `PresenceMotion` and `AtomMotion`?

bsunderhus replied:
Maybe something similar to [griffel fallback properties](https://griffel.js.org/react/api/make-styles#css-fallback-properties)?!

```tsx
const Motion = createPresenceComponent([
  // default motion
  {
    enter: {
      keyframes: [
        { opacity: 0, transform: 'scale(0)' },
        { opacity: 1, transform: 'scale(1)' },
      ],
      duration: 300,
    },
    exit: {
      /* ... */
    },
  }, 
  {/* reduced motion goes here */}
]);
```

robertpenner replied:
@bsunderhus I also think a middle type would be good, to DRY up 
`{ keyframes: Keyframe[] } & KeyframeEffectOptions`.

layershifter replied:
> _from @bsunderhus_ 
> or perhaps we should have a middle type here between `PresenceMotion` and `AtomMotion`?

> _from @robertpenner_ 
> I also think a middle type would be good, to DRY up
> `{ keyframes: Keyframe[] } & KeyframeEffectOptions`.

Do you mean a new type to extract `{ keyframes?: Keyframe[] } & KeyframeEffectOptions`? 

----

> Maybe something similar to [griffel fallback properties](https://griffel.js.org/react/api/make-styles#css-fallback-properties)?!

@bsunderhus I updated the PR description with two options and added some arguments against them. Please check and let me know WDYT.

robertpenner replied:
@layershifter 

> Do you mean a new type to extract { keyframes?: Keyframe[] } & KeyframeEffectOptions?

Is it necessary to make `keyframes` optional in `reducedMotion` (could we look at a use case)? Or would it be simpler to say, "If you're going to make a custom reduced motion, you need to define keyframes"?

layershifter replied:
> Is it necessary to make keyframes optional in reducedMotion (could we look at a use case)? Or would it be simpler to say, "If you're going to make a custom reduced motion, you need to define keyframes"?

I don't think that we need to force consumers there, especially after the duration change (https://github.com/microsoft/fluentui/pull/33353#discussion_r1877844281). Also, the media query itself (https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion) does not put any restrictions.

robertpenner commented:
The feature motivation is solid; the feature design needs to be fleshed out a bit more like an RFC (if it's not going to be an actual RFC). For example: 
- The alternative `reducedMotion` implementation conceivably could be placed in a number of different spots. How do we know this spot is the best design? Comparing alternatives would be instructive.
- How might `reducedMotion` look with other cases, e.g. when `enter` is an array of atoms?

layershifter commented:
> the feature design needs to be fleshed out a bit more like an RFC (if it's not going to be an actual RFC).

FYI, it's was mentioned in the original RFC as a proposal: https://github.com/microsoft/fluentui/blob/master/docs/react-v9/contributing/rfcs/react-components/convergence/motion-definition-n-apis.md#advanced-reduced-motion-support

----

> * The alternative `reducedMotion` implementation conceivably could be placed in a number of different spots. How do we know this spot is the best design? Comparing alternatives would be instructive.

I updated the PR description to list other two options, is there anything that comes to your mind? Did I miss something?

> * How might `reducedMotion` look with other cases, e.g. when `enter` is an array of atoms?

As the atom definition changes, it will work transparently. I added an example to the PR description.

ling1726 commented on packages/react-components/react-motion/library/src/types.ts:
```diff
@@ -1,4 +1,15 @@
+     *
+     * @see https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion
+     */
+    reducedMotion?: { keyframes?: Keyframe[] } & KeyframeEffectOptions;
```
Just throwing out another idea, instead of making the reduced motion options a recursion of the parent motion declaration we use a second argument in the factory and enforce the SRP for motion declarations?

```ts
createPresentionComponent(atom, reducedMotionAtom?)
```

This way we'll avoid TS recursion, and decouple the original motion and its reduced motion alternative. From what I understand even the current `reducedMotion` property completely overrides all other options anyway

robertpenner replied:
It's not recursion though; it's just leaf reuse. There is no atom inside an atom, nor keyframe inside a keyframe.

layershifter replied:
@ling1726 please check "rejected options" in the PR's description, the order issue will be the same here.

ling1726 commented on packages/react-components/react-motion/library/src/hooks/useAnimateAtoms.ts:
```diff
@@ -14,13 +18,19 @@ function useAnimateAtomsInSupportedEnvironment() {
+
           ...(isReducedMotion && { duration: 1 }),
-        });
+          ...(isReducedMotion && reducedMotionParams),
```
If a user has provided reducedMotion params, I think we should stop applying the default `duration: 1` value.

It might sound stupid to pass no duration and it defaults to 0, but we should still the user have the full override power in this case

layershifter replied:
> If a user has provided reducedMotion params, I think we should stop applying the default duration: 1 value.

I wanted to avoid breaks, that's why I kept it. A user will be still able to override `duration`, so I don't see any harm with the current behavior.

robertpenner replied:
@ling1726 @layershifter What if we forced them to override `duration` for `reducedMotion`? After all, if the duration stays at 1 ms, new keyframes aren't going to accomplish much, in most use cases. And there might be someone who adds `reducedMotion` keyframes, then is puzzled when they don't seem to work, because they don't realize the duration is at 1 ms and they need to change it.

layershifter replied:
> @ling1726 @layershifter What if we forced them to override `duration` for `reducedMotion`? After all, if the duration stays at 1 ms, new keyframes aren't going to accomplish much, in most use cases. And there might be someone who adds `reducedMotion` keyframes, then is puzzled when they don't seem to work, because they don't realize the duration is at 1 ms and they need to change it.

That's valid concern. As both you (@robertpenner) and @ling1726 raised it, I updated the implementation:
- `reducedMotion` is `undefined` => `{ duration: 1ms }`
- `reducedMotion` is defined => we will use whatever user provided

ling1726 commented on packages/react-components/react-motion/stories/src/CreateMotionComponent/CreateMotionComponentImperativeRefPlayState.stories.tsx:
```diff
@@ -66,6 +66,10 @@ const FadeEnter = createMotionComponent({
   iterations: Infinity,
+
+  reducedMotion: {
+    iterations: 1,
```
What's the value of having a single iteration here? Is it just that this property has a clearer presence in documentation?

robertpenner replied:
With infinite iterations, the reduced motion with 1 ms was looping and flashing.

layershifter replied:
👆 Exactly, either this or we need to complicate examples significantly.

robertpenner commented:
> > * The alternative `reducedMotion` implementation conceivably could be placed in a number of different spots. How do we know this spot is the best design? Comparing alternatives would be instructive.
> 
> I updated the PR description to list other two options, is there anything that comes to your mind? Did I miss something?

@layershifter I think you've covered it well.

> > * How might `reducedMotion` look with other cases, e.g. when `enter` is an array of atoms?
> 
> As the atom definition changes, it will work transparently. I added an example to the PR description.

Excellent additions, thanks.

robertpenner commented on packages/react-components/react-motion/library/src/types.ts:
```diff
@@ -1,4 +1,15 @@
-export type AtomMotion = { keyframes: Keyframe[] } & KeyframeEffectOptions;
+export type AtomMotion = { keyframes: Keyframe[] } & KeyframeEffectOptions & {
+    /**
+     * Allows to specify a reduced motion version of the animation. If provided, the settings will be used when the
+     * user has enabled the reduced motion setting in the operating system. If not provided, the duration of the
+     * animation will be overridden to be 1ms.
+     *
+     * Note, if keyframes are provided, they will be used instead of the regular keyframes.
+     *
+     * @see https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion
+     */
+    reducedMotion?: { keyframes?: Keyframe[] } & KeyframeEffectOptions;
+  };
```
I don't know what to call it exactly, but factoring out the repeated structure here could improve readability and maintainability:

```suggestion
type AtomCore = { keyframes: Keyframe[] } & KeyframeEffectOptions;

export type AtomMotion = AtomCore & {
  /**
   * Allows to specify a reduced motion version of the animation. If provided, the settings will be used when the
   * user has enabled the reduced motion setting in the operating system. If not provided, the duration of the
   * animation will be overridden to be 1ms.
   *
   * Note, if keyframes are provided, they will be used instead of the regular keyframes.
   *
   * @see https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion
   */
  reducedMotion?: Partial<AtomCore>;
};
```

robertpenner commented on packages/react-components/react-motion/library/src/hooks/useAnimateAtoms.ts:
```diff
@@ -14,13 +18,20 @@ function useAnimateAtomsInSupportedEnvironment() {
-        const animation = element.animate(keyframes, {
-          fill: 'forwards',
-
+        const { keyframes, reducedMotion, ...params } = motion;
+        const { keyframes: reducedMotionKeyframes = keyframes, ...reducedMotionParams } = reducedMotion ?? {
+          duration: 1,
+        };
+
+        const animationKeyframes: Keyframe[] = isReducedMotion ? reducedMotionKeyframes : keyframes;
```
The destructuring and falling back is efficient and concise, but I think it's not that easy to read and follow as-is; it needs to slow down a bit and unpack itself for future readers. In particular, the eyes get a bit dizzy keeping track of the repeated uses of `keyframes`: 
```
const { keyframes ...
const { keyframes: reducedMotionKeyframes = keyframes ...
```

Here's my take on it; perhaps it verges on over-explaining but I think there should be comments to explain the various levels of fallbacks.
```suggestion
        // A motion atom's default reduced motion is a simple 1 ms duration.
        // But an atom can define a custom reduced motion, overriding keyframes and/or params like duration, easing, iterations, etc.
        const defaultReducedMotion: AtomMotion['reducedMotion'] = { duration: 1 };
        // Grab the custom reduced motion definition if it exists, or fall back to the default reduced motion.
        const { keyframes: motionKeyframes, reducedMotion = defaultReducedMotion, ...params } = motion;
        // Grab the reduced motion keyframes if they exist, or fall back to the regular keyframes. 
        const { keyframes: reducedMotionKeyframes = motionKeyframes, ...reducedMotionParams } = reducedMotion;

        const animationKeyframes: Keyframe[] = isReducedMotion ? reducedMotionKeyframes : motionKeyframes;
```

robertpenner commented on packages/react-components/react-motion/library/src/hooks/useAnimateAtoms.ts:
```diff
@@ -14,13 +18,20 @@ function useAnimateAtomsInSupportedEnvironment() {
           ...params,
-          ...(isReducedMotion && { duration: 1 }),
-        });
+
```
```suggestion
          // Use reduced motion overrides (e.g. duration, easing) when reduced motion is enabled.
```
